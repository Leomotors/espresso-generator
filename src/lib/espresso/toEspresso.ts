import { z } from "zod";

function bitsORabcdToBits(bitsORabcd: string, bitsOutput: number) {
  if (bitsORabcd === "-" || !bitsORabcd) {
    return "-".repeat(bitsOutput);
  }

  bitsORabcd = bitsORabcd.toLowerCase();

  if (!bitsORabcd.match(/[a-z]/)) {
    return bitsORabcd.padEnd(bitsOutput, "0");
  }

  return Array.from({ length: bitsOutput }, (_, i) =>
    bitsORabcd.includes(String.fromCharCode(97 + i)) ? "1" : "0",
  ).join("");
}

export function ioToEspresso(
  ioData: Record<string, string>,
  bitsOutput: number,
): [string, string][] {
  return Object.entries(ioData)
    .sort((a, b) => +a[0] - +b[0])
    .map(([key, value]) => [key, bitsORabcdToBits(value, bitsOutput)]);
}

const EspressoSchema = z.array(
  z.tuple([z.string().regex(/^[01]+$/), z.string().regex(/^[01-]+$/)]),
);

export function generateFullEspresso(data: [string, string][]) {
  EspressoSchema.parse(data);

  const bitsInput = data[0][0].length;
  const bitsOutput = data[0][1].length;

  const inputLabel = Array.from({ length: bitsInput }, (_, i) =>
    String.fromCharCode(65 + i),
  ).join(" ");

  const outputLabel = Array.from({ length: bitsOutput }, (_, i) =>
    String.fromCharCode(97 + i),
  ).join(" ");

  return `.i ${bitsInput}
.o ${bitsOutput}
.ilb ${inputLabel}
.ob ${outputLabel}
.p ${data.length}
${data.map(([input, output]) => `${input} ${output}`).join("\n")}
.e`;
}
