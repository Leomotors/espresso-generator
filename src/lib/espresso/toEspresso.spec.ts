import { describe, it, expect } from "vitest";
import { ioToEspresso } from "./toEspresso";

describe("toEspresso", () => {
  it("IO", () => {
    expect(
      ioToEspresso(
        {
          "0000": "ABCDEF",
          "0001": "BC",
          "0010": "ACDEG",
          "0011": "ABCDG",
          "0100": "BCFG",
          "0101": "ACDFG",
          "0110": "ACDEFG",
          "0111": "ABC",
          "1000": "ABCDEFG",
          "1001": "ABCDFG",
          "1010": "-",
          "1011": "-",
          "1100": "-",
          "1101": "-",
          "1110": "-",
          "1111": "-",
        },
        7,
      ),
    ).matchSnapshot();
  });

  it("OI", () => {});
});
