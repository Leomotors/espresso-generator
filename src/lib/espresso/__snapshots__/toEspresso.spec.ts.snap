// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`toEspresso > IO 1`] = `
[
  [
    "0000",
    "1111110",
  ],
  [
    "0001",
    "0110000",
  ],
  [
    "0010",
    "1011101",
  ],
  [
    "0011",
    "1111001",
  ],
  [
    "0100",
    "0110011",
  ],
  [
    "0101",
    "1011011",
  ],
  [
    "0110",
    "1011111",
  ],
  [
    "0111",
    "1110000",
  ],
  [
    "1000",
    "1111111",
  ],
  [
    "1001",
    "1111011",
  ],
  [
    "1010",
    "-------",
  ],
  [
    "1011",
    "-------",
  ],
  [
    "1100",
    "-------",
  ],
  [
    "1101",
    "-------",
  ],
  [
    "1110",
    "-------",
  ],
  [
    "1111",
    "-------",
  ],
]
`;
